packages/block-editor/src/hooks/metadata.js

```diff
@@ -0,0 +1,60 @@
+/**
+ * External dependencies
+ */
+import { has } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { addFilter } from '@wordpress/hooks';
+import { hasBlockMetadataSupport } from '@wordpress/blocks';
+
+/**
+ * Filters registered block settings, extending attributes to include `metadata`.
+ *
+ * see: https://github.com/WordPress/gutenberg/pull/40393/files#r864632012
+ *
+ * @param {Object} blockTypeSettings Original block settings.
+ * @return {Object} Filtered block settings.
+ */
+export function addMetaAttribute( blockTypeSettings ) {
+	// Allow blocks to specify their own attribute definition with default values if needed.
+	if (
+		has( blockTypeSettings.attributes, [
+			'__experimentalMetadata',
```

Comment:
It can be a bit tricky when to use the `__experimental` prefix and when not to — in this case, because the attribute is serialised to post content, if we want to remove the `__experimental` prefix further down the track, we'll need to add either a deprecation or some form of migration. Here's how the markup is currently looking for me:

```html
<!-- wp:group {"__experimentalMetadata":{"name":"My custom name"}} -->
<div class="wp-block-group"></div>
<!-- /wp:group -->
```

I'm wondering if it'd be worth having the block attribute be `metadata` (without the `__experimental` prefix) but have the block support setting (the one in `block.json`) still use the `__experimental` prefix, since changing the `block.json` can happen more easily in code. We appear to be using that approach for the Layout support. Possibly for a similar reason, typically the experimental features in `theme.json` don't appear to use the `__experimental` prefix, so that themes don't need to be updated when a feature is stabilised?

Reply:
There is no precedence with using the experimental prefix for block attributes. You shared a reasonable concern that this name will be serialized in the saved post-type content, which will require further actions. Well, there are ways to mitigate that as well, and we even discussed introducing a filter that would allow dynamically mapping old attribute names to new ones in https://github.com/WordPress/gutenberg/pull/42712. Since this API will be fully controlled by the block editor for start, maybe the `__experimental` prefix can be avoided.

Reply:
Ok I'll remove the __experimental` prefix from the attribute.

Reply:
Removed in https://github.com/WordPress/gutenberg/pull/42605/commits/e6465de97cdb2e0a690b658f93e5e387267182e0